src/AWSCredentials.jl

```diff
@@ -185,263 +189,313 @@ function localhost_is_ec2()
     return false
 end
 
-localhost_maybe_ec2() = localhost_is_ec2() ||
-                        isfile("/sys/devices/virtual/dmi/id/product_uuid")
-
 """
-    aws_user_arn(::AWSConfig)
+    check_credentials(aws_creds::AWSCredentials, force_refresh::Bool)
 
-Unique
-[Amazon Resource Name]
-(http://docs.aws.amazon.com/IAM/latest/UserGuide/id_users.html)
-for configrued user.
+# Arguments
+- `aws_creds::AWSCredentials`: AWSCredentials to be checked / refreshed
+- `force_refresh::Bool`: `true` to refresh the credentials
 
-e.g. `"arn:aws:iam::account-ID-without-hyphens:user/Bob"`
+Checks current AWSCredentials, refreshing them if they are soon to expire. If `force_refresh` is
+`true` the credentials will be renewed immediately
 """
-function aws_user_arn(aws::AWSConfig)
-    creds = aws[:creds]
-    if creds.user_arn == ""
-        r = Services.sts(aws, "GetCallerIdentity", [])
-        creds.user_arn = r["Arn"]
-        creds.account_number = r["Account"]
-    end
-    return creds.user_arn
-end
+function check_credentials(aws_creds::AWSCredentials; force_refresh::Bool=false)
+    if force_refresh || _will_expire(aws_creds)
+        credential_method = aws_creds.renew
 
+        if credential_method !== nothing
+            new_aws_creds = credential_method()
 
-"""
-    aws_account_number(::AWSConfig)
+            new_aws_creds === nothing && error("Can't find AWS credentials!")
+            copyto!(aws_creds, new_aws_creds)
 
-12-digit [AWS Account Number](http://docs.aws.amazon.com/general/latest/gr/acct-identifiers.html).
-"""
-function aws_account_number(aws::AWSConfig)
-    creds = aws[:creds]
-    if creds.account_number == ""
-        aws_user_arn(aws)
+            # Ensure credential_method is not overwritten by the new credentials
+            aws_creds.renew = credential_method
+        end
     end
-    return creds.account_number
+
+    return aws_creds
 end
 
 
-"""
-    ec2_metadata(key)
+function _will_expire(aws_creds::AWSCredentials)
+    return now(UTC) >= aws_creds.expiry - Minute(5)
+end
+
 
-Fetch [EC2 meta-data]
-(http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html)
-for `key`.
 """
-function ec2_metadata(key)
+    _ec2_metadata(metadata_endpoint::String)
 
-    @assert localhost_maybe_ec2()
+# Arguments
+- `metadata_endpoint::String`: AWS internal meta data endpoint to hit
 
-    String(http_get("http://169.254.169.254/latest/meta-data/$key").body)
+Retrieve the EC2 meta data from the local AWS endpoint.
+"""
+function _ec2_metadata(metadata_endpoint::String)
+    try
+        http_request = HTTP.request
+        request = @mock http_request(
```

Comment:
This works if you use Mocking 0.7
```suggestion
        request = @mock HTTP.request(
```

Reply:
Perfect! That'll make things a bit cleaner :)